src/main/haskell/kore/src/Kore/Builtin/KEqual.hs

```diff
@@ -0,0 +1,158 @@
+                 Simplifier )
+import           Kore.Step.StepperAttributes
+                 ( StepperAttributes )
+import           Kore.Unification.Unifier
+                 ( unificationProcedure )
+import           Kore.Unparser.Unparse
+                 ( unparseToString )
+
+
+groundObjectSort :: String -> Kore.Sort Object
+groundObjectSort name =
+    Kore.SortActualSort
+        Kore.SortActual
+        { sortActualName =
+            Kore.Id
+            { getId = name
+            , idLocation = Kore.AstLocationImplicit
+            }
+        , sortActualSorts = []
+        }
+
+
+{- | K Sort
+ -}
+kSort :: Kore.Sort Object
+kSort = groundObjectSort "SortK"
+
+{- | Bool Sort
+ -}
+boolSort :: Kore.Sort Object
+boolSort = groundObjectSort "SortBool"
+
+{- | Verify that the sort is the @SortK@.
+
+ -}
+assertKSort :: Builtin.SortVerifier
+assertKSort _ sort =
+    Kore.Error.koreFailWhen (sort /= kSort)
+        ("Sort '" ++ unparseToString sort ++ "' is not SortK")
+
+{- | Verify that hooked symbol declarations are well-formed.
+
+  See also: 'Builtin.verifySymbol'
+
+ -}
+symbolVerifiers :: Builtin.SymbolVerifiers
+symbolVerifiers =
+    HashMap.fromList
+    [ ("KEQUAL.eq", Builtin.verifySymbol Bool.assertSort [assertKSort, assertKSort])
+    , ("KEQUAL.neq", Builtin.verifySymbol Bool.assertSort [assertKSort, assertKSort])
+    ]
+
+{- | @builtinFunctions@ are builtin functions on the 'Bool' sort.
+ -}
+builtinFunctions :: Map String Builtin.Function
+builtinFunctions =
+    Map.fromList
+    [ ("KEQUAL.eq", keq)
+    , ("KEQUAL.neq", kneq)
+    ]
+
+keq :: Builtin.Function
+keq = ApplicationFunctionEvaluator (evalKEq True False)
+
+kneq :: Builtin.Function
+kneq = ApplicationFunctionEvaluator (evalKEq False True)
+
+evalKEq
+    :: Bool
+    -> Bool
+    -> MetadataTools Object StepperAttributes
+    -> PureMLPatternSimplifier Object Kore.Variable
+    -> Kore.Application Object (PureMLPattern Object Kore.Variable)
+    -> Simplifier
+        ( AttemptedFunction Object Kore.Variable
+        , SimplificationProof Object
+        )
+evalKEq true false tools _ pat =
+    case pat of
+        Kore.Application _ [t1, t2] -> evalEq t1 t2
+        _ -> failedToEval
+  where
+    evalEq t1 t2 =
+        case unificationProcedure tools t1 t2 of
+            Right (subst, predicate, _)
+                | Predicate.isFalse predicate ->
+                    trivialEvalResult (Bool.asPattern boolSort false)
+                | null subst -> trivialEvalResult (Bool.asPattern boolSort true)
+                | otherwise -> failedToEval
+            Left err -> case unificationToStepError () (Left err) of
+                Right () -> trivialEvalResult (Bool.asPattern boolSort false)
+                Left _ -> failedToEval
+
+failedToEval
+    :: Simplifier
+         (AttemptedFunction level1 variable, SimplificationProof level2)
+failedToEval = pure (NotApplicable, SimplificationProof)
+
+trivialEvalResult
```

Comment:
I'm not sure, what is trivial about this result?

Reply:
It's just a pure pattern, already evaluated.